pkg/internal/telemetry/submitter.go

```diff
@@ -0,0 +1,156 @@
+	host                        string
+	apiKey                      string
+	mutex                       *sync.Mutex
+}
+
+type Payload = prefabProto.TelemetryEvents
+
+func NewTelemetrySubmitter(options options.Options) *Submitter {
+	aggregators := []Aggregator{}
+
+	contextAggregator := NewContextAggregator(options)
+
+	if contextAggregator != nil {
+		aggregators = append(aggregators, contextAggregator)
+	}
+
+	var evaluationSummaryAggregator *EvaluationSummaryAggregator
+	if options.CollectEvaluationSummaries {
+		evaluationSummaryAggregator = NewEvaluationSummaryAggregator()
+		aggregators = append(aggregators, evaluationSummaryAggregator)
+	}
+
+	return &Submitter{
+		aggregators:                 aggregators,
+		host:                        options.TelemetryHost,
+		apiKey:                      options.APIKey,
+		contextAggregator:           contextAggregator,
+		evaluationSummaryAggregator: evaluationSummaryAggregator,
+		mutex:                       &sync.Mutex{},
+		instanceHash:                options.InstanceHash,
+	}
+}
+
+func (ts *Submitter) StartPeriodicSubmission(interval time.Duration) {
+	ticker := time.NewTicker(interval)
+
+	go func() {
+		for range ticker.C {
+			if len(ts.aggregators) > 0 {
+				ts.Submit()
+			}
+		}
+	}()
+}
+
+func (ts *Submitter) RecordEvaluation(data internal.ConfigMatch) {
+	if ts.evaluationSummaryAggregator == nil || !data.IsMatch {
+		return
+	}
+
+	// We don't track log level evaluations
+	if _, ok := data.Match.GetType().(*prefabProto.ConfigValue_LogLevel); ok {
+		return
+	}
+
+	ts.evaluationSummaryAggregator.Record(data)
+}
+
+func (ts *Submitter) RecordContext(data *contexts.ContextSet) {
+	if ts.contextAggregator == nil {
+		return
+	}
+
+	ts.contextAggregator.Record(data)
+}
+
+func (ts *Submitter) Submit() error {
+	ts.mutex.Lock()
+	defer ts.mutex.Unlock()
+
+	if len(ts.aggregators) == 0 {
+		return errors.New("no aggregators to submit")
+	}
+
+	payload := Payload{
+		InstanceHash: ts.instanceHash,
+	}
+
+	for _, aggregator := range ts.aggregators {
+		aggregator.Lock()
+
+		data := aggregator.GetData()
+		if data == nil {
+			continue
+		}
+
+		payload.Events = append(payload.Events, data)
+
+		aggregator.Clear()
+		aggregator.Unlock()
+	}
+
+	if len(payload.Events) == 0 {
+		return nil
+	}
+
+	url := fmt.Sprintf("%s/api/v1/telemetry", ts.host)
+
+	payloadData, err := proto.Marshal(&payload)
+	if err != nil {
+		return fmt.Errorf("failed to marshal payload: %v", err)
+	}
+
+	req, err := http.NewRequest("POST", url, bytes.NewReader(payloadData))
+	if err != nil {
+		return fmt.Errorf("failed to create request: %v", err)
+	}
+
+	req.Header.Set("Content-Type", "application/x-protobuf")
+	req.Header.Set("Accept", "application/x-protobuf")
+	req.Header.Set("X-PrefabCloud-Client-Version", internal.ClientVersionHeader)
+
+	encodedAuth := base64.StdEncoding.EncodeToString([]byte("authuser:" + ts.apiKey))
+	req.Header.Set("Authorization", "Basic "+encodedAuth)
+
+	client := &http.Client{}
```

Comment:
is this making a new client for every request? my friend claude says that we should reuse these clients (thread safe), and that they manage connection pools etc behind the scenes

Reply:
Will do, thanks